### ISSUE TITLE
addon list --verbose  should show imagestreams and templates tag/version information

### ISSUE THREAD

**Issue by siamaksade** [minishift/minishift#1479]
Instead of duplicating the imagestreams and templates in the minishift repository, the xpaas addon should import them from the upstream repositories matching the openshift version:
https://github.com/openshift/openshift-ansible/blob/master/roles/openshift_examples/files/examples/v3.6/xpaas-streams/jboss-image-streams.json

**Comment by gbraad** [comment]
This means that the OpenShift version should be included in the addon and we need to download the streams. At the moment we do not do this, as this will increase the testload...

although, it will also guarantee this is in line with the expected oc/openshift version.

@agajdosi @LalatenduMohanty @praveenkumar WDYT

**Comment by praveenkumar** [comment]
Right, Currently we cached all the imagestrem in addon directory and then use it with `oc create -f`. We can use addon variable which default value can be set as per our release and the user can modify as per their need.

```
# Name: xpaas
# Description: Imports xPaaS templates
# Url: https://github.com/openshift/openshift-ansible/tree/master/roles/openshift_examples/files/examples/v3.7
# Required-Vars: OPENSHIFT_VERSION
# Default-Envs: OPENSHIFT_VERSION=v3.6

oc apply -fhttps://github.com/openshift/openshift-ansible/raw/master/roles/openshift_examples/files/examples/#{OPENSHIFT_VERSION}/xpaas-streams/fis-image-streams.json
oc apply -f https://github.com/openshift/openshift-ansible/raw/master/roles/openshift_examples/files/examples/#{OPENSHIFT_VERSION}/xpaas-streams/jboss-image-streams.json -n openshift
```

Defaults-Env feature is targeted for this sprint #1135

**Comment by praveenkumar** [comment]
@minishift/minishift-dev I want your fedback on this.

**Comment by jorgemoralespou** [comment]
This will allow to install new versions, but will not be self-contained. That means that applying the addon without internet connectivity will fail.

But, that's what I do in "oc-cluster". But not what I would do in CDK.

**Comment by jorgemoralespou** [comment]
I think that the metadata is in the Imagestream. Only it will resolve to
the sha256 of the image one it connects, but not to create/load them into
the system.

**Comment by siamaksade** [comment]
Got it.

However I think it's still improbable to assume the user doesn't have
internet connection when using xpaas images. They won't be able to use s2i
from source in that case and they have to have all the maven dependencies
in their local repo if they do binary deploy. Although possible buy my own
experience is that always something is missing and I needed internet.

**Comment by jorgemoralespou** [comment]
But the question is whether enforcing internet connection when installing
the add-on or when using the functionality provided.

In any case, many addons will require internet connection to pull down the
images used, so maybe you're right and this should be enabled.

**Comment by gbraad** [comment]
Internet connection is not per definition a requirement, as the images can
be exported and cached.

Using a fixed version will help with our testing effort as there is an
expectation we can have. With remote resouces we introduce an uncommon
factor... The connection can simply timeout, orworse, the resource is
actually faulty, or even has changed requirements. Will these things
shouldn't happen, they unfortunately do occur.

But yes, being up-to-date is equally important. At least if the version can
be fixed in the add-on, we are in a slightly better situation, as within
the same version no breaking changes should occur. However, some of the
earlier issues still apply.

-- 

   Gerard Braad | http://gbraad.nl
   [ Doing Open Source Matters ]

**Comment by siamaksade** [comment]
Throwing out wild ideas here but is there any possiblity to download
various versions of the imagestream and templates from upstream when
minishift is being packaged, instead of having them duplicated in the
minishift repository?

I don't know the release process for minishift but it's fairly easily to
write a bash script that does this.

The xpaas folder would then look like this:
xpaas
       v3.6.0
              *.json
       v1.5.0
              *.json

The xpaas addon script can also be modified to first check for the cached
image streams and if they don't exist, download from upstream.

This would solve the test dependency issue and also allow matching image
stream versions  with OpenShift.

**Comment by hferentschik** [comment]
> This will allow to install new versions, but will not be self-contained. 

+1

@siamaksade What exact problem do you try to solve? Did you have a concrete problem with the current approach? You can always create and use your own add-on which imports via a URL.

**Comment by siamaksade** [comment]
@hferentschik the exact problem is that when I used minishift v1.6.0 to start OpenShift v3.6.0, the imagestreams installed by the xpaas addon do not match the version of OpenShift. For example `jboss-webserver31-tomcat8-openshift:1.0` is missing. 

Therefore as a user, I can not trust the xpaas addon to install the correct image streams for the OpenShift version I have chosen and I have to always import the xpaas image streams myself if I want to be sure my OpenShift cluster has the correct imagestreams.

This problem will always exist as long as minishift uses a single version of imagestreams for all OpenShift versions. 

Minishift still can be self-contained if the xpaas addon distinguishes between versions of OpenShift, similar to this repo structure:
https://github.com/openshift/openshift-ansible/tree/master/roles/openshift_examples/files/examples

However since a specific version of minishift can be used to start OpenShift versions that are released after minishift, there is still a need to pull them from upstream if the cached ones don't match the OpenShift version.

**Comment by gbraad** [comment]
> when I used minishift v1.6.0 to start OpenShift v3.6.0, the imagestreams installed by the xpaas addon do not match the version of OpenShift.

understandable, so it would be nice to have OPENSHIFT_VERSION being populated with the version that Minishift actually installs. And then use a set of stored files to resolve them on disk.

**Comment by siamaksade** [comment]
Exactly. And modify the xpaas addon to import the image streams from
upstream (openshift repo) if it doesn't find them on the disk for the
specified version.

**Comment by gbraad** [comment]
I see several edge cases and additional logic, so I think we need to look
into this; needs investigation.

Maybe something is possible with the minimal open shift flag, perhaps
conditions. It is a straightforward implementation or template change.

-- 

   Gerard Braad | http://gbraad.nl
   [ Doing Open Source Matters ]

**Comment by jorgemoralespou** [comment]
@siamaksade this is not a problem of the addon, this is a problem of the
templates and imagestreams. They do use dockerImageRepoitory and not a from
in every IST. There's an issue opened for some time but that team has not
fixed it.
If you use the 3.7 tag or any other will have the same problem. This is
currently broken for every OpenShift install in the world. We have patched
it ourselves for online to avoid the lack of response from that team.
Probably something that needs to be managed in a different manner. Let's
talk on Monday, and discuss with Diógenes.

**Comment by siamaksade** [comment]
@jorgemoralespou that's a different issue that coincides with this one. 

The problem I referred to was causes due to lack of the image stream `jboss-webserver31-tomcat8-openshift:1.0` in the imagestream json file on disk:
https://raw.githubusercontent.com/minishift/minishift/7a7156589a8d078d436ca9d63c1aced97456058a/addons/xpaas/xpaas-streams/jboss-image-streams.json

**Comment by jorgemoralespou** [comment]
@siamaksade, which is correct since that version is not present in 1.5 or
3.6 version of the Imagestream.

**Comment by siamaksade** [comment]
It is available in 3.6:
https://github.com/openshift/openshift-ansible/blob/master/roles/openshift_examples/files/examples/v3.6/xpaas-streams/jboss-image-streams.json#L149

**Comment by jorgemoralespou** [comment]
@siamaksade right. There's a mess of versions. But that one is also
available in 1.5.

https://github.com/openshift/openshift-ansible/blob/master/roles/openshift_examples/files/examples/v1.5/xpaas-streams/jboss-image-streams.json#L149


I still don't see where is the difference between the Imagestream in 1.5 or
3.6 with respect to this Imagestream.

Maybe I should download them all and diff them. I'll try tomorrow.

El 8 oct. 2017 22:10, "Siamak Sadeghianfar" <notifications@github.com>
escribió:

It is available in 3.6:
https://github.com/openshift/openshift-ansible/blob/master/
roles/openshift_examples/files/examples/v3.6/xpaas-
streams/jboss-image-streams.json#L149

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/minishift/minishift/issues/1479#issuecomment-335034451>,
or mute the thread

**Comment by siamaksade** [comment]
You are comparing the wrong files :) Compare it to the json file in the xpaas addons folder when minishift 1.6 was released:
https://github.com/minishift/minishift/blob/7a7156589a8d078d436ca9d63c1aced97456058a/addons/xpaas/xpaas-streams/jboss-image-streams.json

**Comment by jorgemoralespou** [comment]
@siamaksade, I see. Way too old.

**Comment by praveenkumar** [comment]
I did update this recently, coming release (1.7.0) have it in. But we still need a better solution to this problem. As per discussion above we can have a bash script which will make sure before release it will always update images streams but then we are shipping something as default which not even tested once. Not to mention we are using same for downstream. As long as ansible examples are well tested with the respective openshfit version, we can make it updated all the time with scripts.

**Comment by LalatenduMohanty** [comment]
Personally I am not in favor or downloading on the fly as it creates a lot of unknown variables. Mostly because I am not aware what kind of tests the respective upstream runs to validate the image streams. If I come to know that the image streams are fairly tested for each Origin release my confidence will improve and then we might suggest providing an alternate way where user can update the image-stream on the fly if user wants. In that case it should be a script and should be part of the add-on and if users provides a variable e.g. download-from-upstream=true then it should download the image streams.   

However we should create a script which we can run for each release and see anything need update. How to plugin the script in the CI is a question which I do not have the answer right now.

**Comment by chargio** [comment]
What about oc template-default upgrade?

I have found that the ruby version in minishift 1.7.0 does not include ruby 2.4, while openshift online does.

**Comment by gbraad** [comment]
discrepancies can always exist, and we try to match as close as possible. however, you can add your own imagestreams and templates... and if you feel this needs to be shared, create an addon for https://github.com/minishift/minishift-addons

> That could make things work, as it could test connectivity before updating the templates

what is 'that'?

> it could test connectivity before updating the templates

In some cases download is not prefered as we could be behind a proxy/firewall, and the local definition at least will be applied, and images can be restored from image cache.

**Comment by chargio** [comment]
Using a command line like `oc template defaults upgrade` is what would make things work.
You need to actively use it, so if you are behind a firewall or just not connected to the internet you know that it can't work.
It allows meaningful feedback: "it was not possible to connect to the upstream repo", "there was a mistake in the yaml definitions", "these are not the droids you are looking for"

**Comment by gbraad** [comment]
```
$ eval `minishift oc-env`
$ oc template defaults upgrade
Error: unknown command "template" for "oc"
Run 'oc --help' for usage.
$ oc version
oc v3.6.0+c4dd4cf
kubernetes v1.6.1+5115d708d7
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://192.168.42.121:8443
openshift v3.6.0+c4dd4cf
kubernetes v1.6.1+5115d708d7
```

**Comment by jorgemoralespou** [comment]
Quick note. Moving forward we have a source for curating version of
imagestreams and templates. This is openshift/library GitHub repo. We
should switch to get the imagestreams and templates from here.

**Comment by gbraad** [comment]
@jorgemoralespou good news!

**Comment by praveenkumar** [comment]
@jorgemoralespou should we make a switch now to https://github.com/openshift/library/tree/master/official/eap now?

**Comment by jorgemoralespou** [comment]
Yes.

**Comment by praveenkumar** [comment]
@jorgemoralespou there is no tag in this  https://github.com/openshift/library/tree/master/official/eap repo like we have for ansible one https://github.com/openshift/openshift-ansible/tree/release-3.7  so should we always go with master now?

**Comment by jorgemoralespou** [comment]
I think so.

Cc/ @bparees

**Comment by bparees** [comment]
i'm not going to read through this entire thread so i don't know exactly what i'm being asked, but the "library" does not have release specific branches.  So whatever is in there is only guaranteed to work with the current openshift release (if even that is guaranteed).

Consumers of the library should be managing their own copies of the content they want to redistribute, in a way that ensures they are distributing a copy of the content that is compatible with the openshift cluster version they are distributing.

**Comment by jorgemoralespou** [comment]
@praveenkumar I guess that since minishift needs to distribute the
imagestreams for the release it embes, it will need to look back to
OpenShift ansible repository.

I'm sorry for the noise. I was confused about this repository.

**Comment by praveenkumar** [comment]
> it will need to look back to OpenShift ansible repository. 

Thanks @jorgemoralespou for confirmation. So whatever we have right now in the good state.

**Comment by praveenkumar** [comment]
I am closing this issue as per https://github.com/minishift/minishift/issues/1479#issuecomment-350668951 comment.

**Comment by LalatenduMohanty** [comment]
> If a user uses a different version of Openshift then default then he/she should pull down respective version templates.

@praveenkumar But thats not convenient at all for the user at all.

**Comment by LalatenduMohanty** [comment]
```
$ minishift addon list --verbose
Name       : anyuid
Description: Changes the default security context constraints to allow pods to run as any user.
Enabled    : enabled
Priority   : 0

Name       : admin-user
Description: Create admin user and assign the cluster-admin role to it.
Enabled    : disabled
Priority   : 0

Name       : che
Description: Setup and Configure Eclipse Che Template and Image Streams
Enabled    : disabled
Priority   : 0

Name       : registry-route
Description: Create an edge terminated route for the OpenShift registry
Enabled    : disabled
Priority   : 0

Name       : xpaas
Description: Imports xPaaS templates
Enabled    : disabled
Priority   : 0
```

**Comment by praveenkumar** [comment]
As per discussion, we will tag specific version Openshift for template and image-stream and also add URL to the description so the user can get info when using verbose.

**Comment by LalatenduMohanty** [comment]
Resolved via #2521
